src/puppetlabs/puppetdb/http/command.clj

```diff
@@ -77,40 +92,145 @@
+    (let [req (if-let [cmd (params "command")]
+                (assoc-in req [:params "command"] (str/replace cmd "_" " "))
+                req)
+          req (if-let [ver (params "version")]
+                (assoc-in req [:params "version"] (Integer/parseInt ver))
+                req)
+          {:strs [command version certname]} params
+          missing-items (fn [kind command version certname]
+                          (http/error-response
+                           (str "Missing POST " kind ":"
+                                (when-not command " command")
+                                (when-not version " version")
+                                (when-not certname " certname"))))]
+      (cond
+        (and command version certname) (handle req)
+
+        (or command version certname)
+        (missing-items "parameters" command version certname)
+        
+        :else        
+        (do
+          (log/warn
+           (str "Unable to stream command posted without parameters"
+                " (loading into RAM)"))
+          (if-let [body-str (request/body-string req)]
+            (let [body (json/parse-string body-str)                    
+                  required-fields #{"command" "version" "payload"}]
+              (if-not (= required-fields (set (keys body)))
+                (http/error-response
+                 (str "Unexpected POST command fields: "
+                      (str/join " " (set/difference (set (keys body))
+                                                    required-fields))))
+                (let [certname (get-in body ["payload" "certname"])
+                      params (merge params
+                                    (select-keys body ["command" "version"])
+                                    (when certname {"certname" certname}))
+                      {:strs [command version certname]} params]
+                  ;; Enforce certname too? (currently breaks tests)
+                  (if-not (and command version)
+                    (missing-items "elements" command version true)
+                    (handle
+                     (assoc req
+                            :body (json/generate-string (body "payload"))
+                            :params (assoc params "version"
+                                           (Integer. version))))))))
+            (http/error-response "Empty application/json POST body")))))))
+
+(defn- realize-body [body reject-large-commands? max-command-size]
+  "Returns the body as an in-memory string or byte-array, reading it
+  if necessary.  Throws ::body-stream-overflow if the max-command-size
+  is not respected."
+  (if-not reject-large-commands?
+    (if (instance? java.io.InputStream body)
+      (IOUtils/toByteArray body)
+      body)
+    (cond
+      (instance? java.io.InputStream body)
+      (IOUtils/toByteArray
+       (restrained-drained-stream body (long max-command-size)))
+
+      (string? body) ;; Assuming UCS-2
```

Comment:
I get what you're doing here, but maybe we should say it a bit more clearly in the comment